### PR TITLE
Add V3.0.2 Changes to 3.1.0

### PR DESCRIPTION
**Pull request by sebbader-sap**

### CODE REVIEW

**Comment by sebbader-sap** [comment]
Shall be merged after #254 (I just noticed I had the wrong source branch...)

File: Part2-API-Schemas/openapi.yaml [hunk 1]
```diff
@@ -822,14 +822,9 @@ components:
...
-      properties:
-        submodelElements:
-          type: array
-          items:
-            $ref: 'https://api.swaggerhub.com/domains/Plattform_i40/Part1-MetaModel-Schemas/V3.0.2#/components/schemas/SubmodelElement'
-          minItems: 1
+      $ref: "#/components/schemas/ValueOnly"
     ValueOnly: 
-      description: "The ValueOnly serialization of submodel elements (patternProperties and propertyNames will be supported propably with OpenApi 3.1). The full description of the generic JSON validation schema for the ValueOnly-serialization can be found in chapter 11.4.3 in Details of the Asset Administration Shell Part 2.\n"
+      description: "The ValueOnly serialization (patternProperties and propertyNames will be supported probably with OpenApi 3.1). The full description of the generic JSON validation schema for the ValueOnly-serialization can be found in chapter 11.4 in 'Specification of the Asset Administration Shell - Part 2'."
```
Comment (BirgitBoss) [inline Part2-API-Schemas/openapi.yaml:827]: There are no chapter numbers in asciidoc any longer

File: documentation/IDTA-01002-3/modules/ROOT/pages/IDTA-01002_HTTP-REST-API.adoc [hunk 2]
```diff
@@ -279,7 +281,12 @@ h|Operation Name h|HTTP Verb h|REST-Path h|Comment (e.g. optional query paramete
 |GetAssetAdministrationShell |GET |/aas |content-suffix: $reference
 |PutAssetAdministrationShell |PUT | |
 |GetAllSubmodelReferences |GET |/aas/submodel-refs |Pagination
-|PostSubmodelReference |POST |/aas/submodel-refs |
+|PostSubmodelReference |POST |/aas/submodel-refs | Location header of the response contains the value ‘<baseUrl>/aas/submodel-ref/{submodelIdentifier}’
```
Comment (BirgitBoss) [inline documentation/IDTA-01002-3/modules/ROOT/pages/IDTA-01002_HTTP-REST-API.adoc:284]:
```suggestion
|PostSubmodelReference |POST |/aas/submodel-refs a| Location header of the response contains the value ‘<baseUrl>/aas/submodel-ref/{submodelIdentifier}’ 
```

File: documentation/IDTA-01002-3/modules/ROOT/pages/IDTA-01002_HTTP-REST-API.adoc [hunk 3]
```diff
@@ -255,8 +255,10 @@ GET /submodel/$value?level=deep&extent=withBlobValue
 
 The following constraints apply for the combination of modifiers:
 
+* For Content=Value, the requested object is always serialized to an unnamed JSON Object or Array. This means that the response object must not have with the object’s idShort at the root level.
```
Comment (mjacoby) [inline documentation/IDTA-01002-3/modules/ROOT/pages/IDTA-01002_HTTP-REST-API.adoc:258]:
```suggestion
* For Content=Value, the requested object is always serialized to an unnamed JSON Object or Array. This means that the response object must not have a property with the object’s idShort at the root level.
```

File: documentation/IDTA-01002-3/modules/ROOT/pages/IDTA-01002_HTTP-REST-API.adoc [hunk 4]
```diff
@@ -255,8 +255,10 @@ GET /submodel/$value?level=deep&extent=withBlobValue
...
 * If Level=Core and Content=Value, only the requested object and the direct children without their value (empty value) will be returned in value serialization. If a direct child is a SubmodelElementCollection, "<SubmodelElementCollection/idShort>": \{} will be returned. If a direct child is a SubmodelElementList, "<SubmodelElementList/idShort>": [] will be returned.
 * The combination of Content=Metadata and Extent=WithBLOBValue is not allowed.
+* The behavior of Level in combination with Content=Metadata is undefined. A server may ignore the Level modifier if it is set.
```
Comment (mjacoby) [inline documentation/IDTA-01002-3/modules/ROOT/pages/IDTA-01002_HTTP-REST-API.adoc:261]: What is the alternative to ignoring it? Returning `404 - Not Found` is not allow, right? And actually applying the level modifier is also not possible. 
This statement should clearly list the possible allowed behavior.

File: documentation/IDTA-01002-3/modules/ROOT/pages/IDTA-01002_Spec_InterfacesPayload.adoc [hunk 5]
```diff
@@ -260,14 +260,14 @@ e|https://admin-shell.io/aas/API/3/0/SubmodelServiceSpecification/SSP-001 |Indic
...
+e|https://admin-shell.io/aas/API/3/1/SubmodelRegistryServiceSpecification/SSP-003 |Indicates that the server implemented all details of the Submodel Registry Service Specification Bulk Profile in version 3.1.
 e|https://admin-shell.io/aas/API/3/0/DiscoveryServiceSpecification/SSP-001 |Indicates that the server implemented all details of the Discovery Service Specification Full Profile in version 3.0.
-e|https://admin-shell.io/aas/API/3/0/RepositoryServiceSpecification/SSP-001 |Indicates that the server implemented all details of the Repository Service Specification Full Profile in version 3.0.
-e|https://admin-shell.io/aas/API/3/0/RepositoryServiceSpecification/SSP-002 |Indicates that the server implemented all details of the Repository Service Specification Read Profile in version 3.0.
+e|https://admin-shell.io/aas/API/3/1/DiscoveryServiceSpecification/SSP-002 |Indicates that the server implemented all details of the Discovery Service Specification Full Profile in version 3.0.
```
Comment (mjacoby) [inline documentation/IDTA-01002-3/modules/ROOT/pages/IDTA-01002_Spec_InterfacesPayload.adoc:270]: - URL says it is version 3.1, text says v3.0 - probably text is wrong
- For `DiscoveryServiceSpecification` there is on URL ending in **SSP-001** and one ending in **SSP-002** while both refer to the `Full` profile, although in different versions (v3.0 vs v3.1). Is this intentional? I would expect for the number not to change with versions. Therefore I guess either then `SSP-00x` number is wrong for one of them or the textual description. See https://github.com/admin-shell-io/aas-specs-api/blob/a72897f9faaddf6129951b6822c5c030e6076f15/documentation/IDTA-01002-3/modules/ROOT/pages/IDTA-01002_Spec_InterfacesPayload.adoc?plain=1#L269

**Review by mjacoby — CHANGES_REQUESTED** [review]
For some issues I am not able to decide if they are actual problems, especially the ones related to formatting as there is no way of seeing the final rendered AsciiDoc document (AFAIK).
Some other issues are not absolutely essential but the specification would greatly benefit, e.g. by reducing complexity where it is not helpful, of course without changing the behavior.

File: documentation/IDTA-01002-3/modules/ROOT/pages/IDTA-01002_HTTP-REST-API.adoc (hunk 4 above)
Comment (sebbader-sap) [reply documentation/IDTA-01002-3/modules/ROOT/pages/IDTA-01002_HTTP-REST-API.adoc:261]: What's your suggestion? `400`? `404` doesn't seem correct to me, as the resource in general exists at the specified location.

File: documentation/IDTA-01002-3/modules/ROOT/pages/IDTA-01002_Spec_InterfacesPayload.adoc (hunk 5 above)
Comment (sebbader-sap) [reply documentation/IDTA-01002-3/modules/ROOT/pages/IDTA-01002_Spec_InterfacesPayload.adoc:270]: Explanation of SSP-002 is not correct, I will fix it. The different versions are intended, as SSP-002 will be added with V3.1.0 (and therefore is also not part of the V3.0.2 Word document).

File: documentation/IDTA-01002-3/modules/ROOT/pages/IDTA-01002_HTTP-REST-API.adoc (hunk 4 above)
Comment (mjacoby) [reply documentation/IDTA-01002-3/modules/ROOT/pages/IDTA-01002_HTTP-REST-API.adoc:261]: Well that depends on you understanding of what is a breaking change...

**TL;DR**
Formulate it similar to all other constraints, i.e. _The combination of Content=Metadata and Level is not allowed_

**Detailed Answer**
Right now, the specification is contradictory saying that you can combine metadata and level on the one hand, and on the other hand saying that for metadata serialization there are no properties like _values_ or _submodelElements_.
Therefore I would argue that we can more or less make any change we like in a bugfix version as it will never break a client implementation as they currently cannot know what to expect.
Following that reasoning, I would treat it like any similar constraint, e.g. the one above in the text (The combination of Content=Metadata and Extent=WithBLOBValue is not allowed) and just formulate it the same way.
However, this still leaves room for interpretation what _not allowed_ means. I would argue that it heavily implies returning `400 Bad Request`. But I would not mention this only for this specific case but either on a higher level (applying to all constraints on modifier combinations) or not at all.

File: documentation/IDTA-01002-3/modules/ROOT/pages/Annex/IDTA-01002_SerializationModifierExamples.adoc [hunk 6]
```diff
@@ -205,11 +167,26 @@ If applied to the Submodel:
 [source,json,linenums]
 ----
 {
-  "TechnicalData": {
```
[the remark this answers is not in the document]
Comment (sebbader-sap) [reply documentation/IDTA-01002-3/modules/ROOT/pages/Annex/IDTA-01002_SerializationModifierExamples.adoc:208]: I think I found the right ASCIIDOC command now. Please have a check again.

File: documentation/IDTA-01002-3/modules/ROOT/pages/Annex/IDTA-01002_SerializationModifierExamples.adoc [hunk 7]
```diff
@@ -453,10 +467,8 @@ If applied to the Submodel:
...
-    "RotationSpeed": {
-      "MaxRotationSpeed": "5000"
-    }
+  "RotationSpeed": {
```
[the remark this answers is not in the document]
Comment (sebbader-sap) [reply documentation/IDTA-01002-3/modules/ROOT/pages/Annex/IDTA-01002_SerializationModifierExamples.adoc:496]: I think due to the different "assumptions" before (which mostly are corrected with the V3.0.2 bugfix), they are now pretty much the same. But looked slightly different before...

[the remark this answers is not in the document]
Comment (sebbader-sap) [reply documentation/IDTA-01002-3/modules/ROOT/pages/Annex/IDTA-01002_SerializationModifierExamples.adoc:496]: However, as V3.0.2 is a "only" a bugfix, I don't want to remove the PATCH table. We may want to do so for V3.1.0 but then let's have an own issue/PR for it. Too confusing for me already to keep the Word and ASCIIDOC in sync.

File: documentation/IDTA-01002-3/modules/ROOT/pages/Annex/IDTA-01002_SerializationModifierExamples.adoc (hunk 6 above)
[the remark this answers is not in the document]
Comment (mjacoby) [reply documentation/IDTA-01002-3/modules/ROOT/pages/Annex/IDTA-01002_SerializationModifierExamples.adoc:208]: Formatting looks good now

File: documentation/IDTA-01002-3/modules/ROOT/pages/IDTA-01002_HTTP-REST-API.adoc (hunk 4 above)
Comment (BirgitBoss) [reply documentation/IDTA-01002-3/modules/ROOT/pages/IDTA-01002_HTTP-REST-API.adoc:261]: We should use terms conformant to https://rwth-iat.github.io/aas-specs-antora/IDTA-01001/snapshot/IDTA-01001_Preamble.html#_conformance
i.e. for example:
If parameter Content is set to "Metadata" then Level shall not be defined.

File: documentation/IDTA-01002-3/modules/ROOT/pages/Annex/IDTA-01002_SerializationModifierExamples.adoc [hunk 8]
```diff
@@ -284,7 +287,18 @@ If applied to the Submodel:
...
 ----
 [
-  "RotationSpeed"
+  "TechnicalData"
+  "TechnicalData.RotationSpeed"
+]
+
+----
+
+If applied to the Property inside the SubmodelElementCollection:
```
[the remark this answers is not in the document]
Comment (mjacoby) [reply documentation/IDTA-01002-3/modules/ROOT/pages/Annex/IDTA-01002_SerializationModifierExamples.adoc]: The examples for Path serialization seem to be incorrect. Currently, it says the following for both `depth=deep` and `depth=core`

> If applied to the SubmodelElementCollection:
> 
> [
>   "RotationSpeed.MaxRotationSpeed"
> ]
> 
> If applied to the Property inside the SubmodelElementCollection:
> 
> [
>   "RotationSpeed.MaxRotationSpeed"
> ]


I would have expected the following (maybe I'm wrong here and don't understand how Path serialization is expected to work, but the current status seems inconsistent in itself as it produces the same result when called in the SubmodelElementCollection as well as the Property inside of it).
I am also not 100% if the requested element is itself part of the result. According to https://github.com/admin-shell-io/aas-specs-api/issues/246#issuecomment-1991312483 this is to be decided, however, latest PR to aas-spec v3.0.2 https://github.com/admin-shell-io/aas-specs/pull/372 contains an example where this is the case. So for this proposal I am going to assume it is included. If this is not the case, examples need to be updated in this regard.

As the depth of the nested AAS structure is <= 1 in the example, results are the same for both `depth=deep` and `depth=core`


If applied to the SubmodelElementCollection:
```json
[
  "RotationSpeed"
  "RotationSpeed.MaxRotationSpeed"
]
```

If applied to the Property inside the SubmodelElementCollection:
```json
[
  "RotationSpeed.MaxRotationSpeed"
]
```

File: Part2-API-Schemas/openapi.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (sebbader-sap) [reply Part2-API-Schemas/openapi.yaml]: The formulation has been relaxed slightly, reference to the ASCIIDOC-published HTML will come with V3.1.

File: documentation/IDTA-01002-3/modules/ROOT/pages/Annex/IDTA-01002_SerializationModifierExamples.adoc [hunk 9]
```diff
@@ -289,12 +319,22 @@ If applied to the Submodel:
...
+[
+  "RotationSpeed.MaxRotationSpeed"
+]
+----
 
 If applied to the Property inside the SubmodelElementCollection:
 
 [source,json,linenums]
 ----
-[ ]
```
Comment (BirgitBoss) [inline documentation/IDTA-01002-3/modules/ROOT/pages/Annex/IDTA-01002_SerializationModifierExamples.adoc:297]: Since property has no subelement it seems correct to me to return empty set? Or just return MaxRotationSpeed - but for collection you also do not return reference to collection only, but for its elements only? For Collection you also do not mention the submodel the collection is part of.

File: documentation/IDTA-01002-3/modules/ROOT/pages/IDTA-01002_HTTP-REST-API.adoc [hunk 10]
```diff
@@ -255,8 +255,10 @@ GET /submodel/$value?level=deep&extent=withBlobValue
 
 The following constraints apply for the combination of modifiers:
 
+* For Content=Value, the requested object is always serialized to an unnamed JSON Object or Array. This means that the response object must not have a property with the object’s idShort at the root level.
```
Comment (BirgitBoss) [inline documentation/IDTA-01002-3/modules/ROOT/pages/IDTA-01002_HTTP-REST-API.adoc:258]:
```suggestion
* For Content=Value, the requested object shall always be serialized to an unnamed JSON Object or Array. This means that the response object must not have a property with the object’s idShort at the root level.
```

File: documentation/IDTA-01002-3/modules/ROOT/pages/IDTA-01002_Spec_InterfacesPayload.adoc [hunk 11]
```diff
@@ -260,14 +260,13 @@ e|https://admin-shell.io/aas/API/3/0/SubmodelServiceSpecification/SSP-001 |Indic
...
-e|https://admin-shell.io/aas/API/3/0/RegistryServiceSpecification/SSP-002 |Indicates that the server implemented all details of the Asset Administration Shell Registry Service Specification Full Profile in version 3.0.
-e|https://admin-shell.io/aas/API/3/0/RegistryServiceSpecification/SSP-003 |Indicates that the server implemented all details of the Asset Administration Shell Registry Service Specification Read Profile in version 3.0.
-e|https://admin-shell.io/aas/API/3/0/RegistryServiceSpecification/SSP-004 |Indicates that the server implemented all details of the Submodel Registry Service Specification Full Profile in version 3.0.
-e|https://admin-shell.io/aas/API/3/0/RegistryServiceSpecification/SSP-005 |Indicates that the server implemented all details of the Submodel Registry Service Specification Read Profile in version 3.0.
+e|https://admin-shell.io/aas/API/3/0/AssetAdministrationShellRegistryServiceSpecification/SSP-001 |Indicates that the server implemented all details of the Asset Administration Shell Registry Service Specification Full Profile in version 3.0.
```
Comment (BirgitBoss) [inline documentation/IDTA-01002-3/modules/ROOT/pages/IDTA-01002_Spec_InterfacesPayload.adoc:263]: why is it mixed: sometimes /3/0, sometimes /3/1: this is /3/1, no?

File: documentation/IDTA-01002-3/modules/ROOT/pages/IDTA-01002_Spec_InterfacesPayload.adoc [hunk 12]
```diff
@@ -260,14 +260,13 @@ e|https://admin-shell.io/aas/API/3/0/SubmodelServiceSpecification/SSP-001 |Indic
...
+e|https://admin-shell.io/aas/API/3/0/AssetAdministrationShellRegistryServiceSpecification/SSP-002 |Indicates that the server implemented all details of the Asset Administration Shell Registry Service Specification Read Profile in version 3.0.
+e|https://admin-shell.io/aas/API/3/1/AssetAdministrationShellRegistryServiceSpecification/SSP-003 |Indicates that the server implemented all details of the Asset Administration Shell Registry Service Specification Bulk Profile in version 3.1.
+e|https://admin-shell.io/aas/API/3/0/SubmodelRegistryServiceSpecification/SSP-001 |Indicates that the server implemented all details of the Submodel Registry Service Specification Full Profile in version 3.0.
+e|https://admin-shell.io/aas/API/3/0/SubmodelRegistryServiceSpecification/SSP-002 |Indicates that the server implemented all details of the Submodel Registry Service Specification Read Profile in version 3.0.
+e|https://admin-shell.io/aas/API/3/1/SubmodelRegistryServiceSpecification/SSP-003 |Indicates that the server implemented all details of the Submodel Registry Service Specification Bulk Profile in version 3.1.
```
Comment (BirgitBoss) [inline documentation/IDTA-01002-3/modules/ROOT/pages/IDTA-01002_Spec_InterfacesPayload.adoc:268]: add text somewhere explaining that a server can implement different versions of different service specifications?

File: documentation/IDTA-01002-3/modules/ROOT/pages/Annex/IDTA-01002_SerializationModifierExamples.adoc (hunk 9 above)
Comment (sebbader-sap) [reply documentation/IDTA-01002-3/modules/ROOT/pages/Annex/IDTA-01002_SerializationModifierExamples.adoc:297]: > but for collection you also do not return reference to collection only, but for its elements only?

Yes, as "core" says the thing itself (here the SubmodelElementCollection) + its direct children (the Property). If there were even more levels (additional SubmodelElementCollections or SubmodelElementLists), their elements would be skipped. But here we only have one.

> For Collection you also do not mention the submodel the collection is part of.

Correct, as the IdShortPath needs to start at the first SubmodelElement, not containing the idShort of the Submodel at all.

File: documentation/IDTA-01002-3/modules/ROOT/pages/IDTA-01002_Spec_InterfacesPayload.adoc (hunk 12 above)
Comment (sebbader-sap) [reply documentation/IDTA-01002-3/modules/ROOT/pages/IDTA-01002_Spec_InterfacesPayload.adoc:268]: I created a new task to explain it in V3.1: https://github.com/admin-shell-io/aas-specs-api/issues/264

**Comment by sebbader-sap** [comment]
> questions and minor change proposals

I think I responded to all your comments, please check again.

File: documentation/IDTA-01002-3/modules/ROOT/pages/Annex/IDTA-01002_SerializationModifierExamples.adoc (hunk 8 above)
[the remark this answers is not in the document]
Comment (sebbader-sap) [reply documentation/IDTA-01002-3/modules/ROOT/pages/Annex/IDTA-01002_SerializationModifierExamples.adoc]: Thanks for pointing me again to this topic, have changed it now.

**Review by mjacoby — APPROVED** [review]
